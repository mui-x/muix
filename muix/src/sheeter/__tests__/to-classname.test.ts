import { initPlatform, dump } from "./init-platform.t"
import { $W, $T, $V, $I, V, T, I, TTyped } from 'reactxx-typings'

import { getTypedEngine } from 'reactxx-sheeter'

interface Shape {
  theme: { primary }
  sheetQuery: { opened: boolean }
}

const { $themed, $if, $ifelse, $web, $native, $rules, $toClassNames, $atomize
} = getTypedEngine<Shape>()

describe("TO CLASSNAMES", () => {

  type Par = { opened: boolean, theme: null }
  const sheetQuery = (opened: boolean) => ({ $sheetQuery: { opened } })
  const query = (opened: boolean, ...ruleset: TTyped.RulesetSimple[]) => dump($toClassNames<TTyped.RulesetIds>(sheetQuery(opened), ...ruleset))

  const doTest = (isWeb: boolean) => {
    beforeEach(() => initPlatform(isWeb))
    let ruleset

    it("01: null", () => dump($toClassNames(null, null)))
    it("02: null, {}, undefined", () => dump($toClassNames(null, undefined)))
    it("03: mixed", () => dump($toClassNames(null,
      $atomize<T>({ color: 'red' }), // ruleset source
      $atomize<T>({ color: 'blue' }, { color: 'yellow' }), // atomized ruleset
      $if<T>(p => true, { color: 'maroon' }), // temporary item
      $toClassNames<T>(null, $atomize<T>({ color: 'green' })) // atomized ruleset
    )))
    it("04: query => opened", () => query(true,
      $if(p => p.$sheetQuery.opened, { color: 'red' }),
      $if(p => !p.$sheetQuery.opened, { color: 'blue' }),
    ))
    it("05: query => closed", () => query(false,
      $if(p => p.$sheetQuery.opened, { color: 'red' }),
      $if(p => !p.$sheetQuery.opened, { color: 'blue' }),
    ))
    it("06: query => closed => pre-atomized", () => query(false, $atomize<T>(
      $if<T>(p => p.$sheetQuery.opened, { color: 'red' }),
      $if<T>(p => !p.$sheetQuery.opened, { color: 'blue' }),
    )))
    it("07: query => closed => pre-queried", () => query(false,
      $toClassNames<T>(sheetQuery(false),
        $if<T>(p => p.$sheetQuery.opened, { color: 'red' }),
        $if<T>(p => !p.$sheetQuery.opened, { color: 'blue' }),
      )
    ))
    it("08: static if and ifelse", () => {
      const inRender = (disabled: boolean) =>
        dump($toClassNames<T>(sheetQuery(false),
          $if<T>(disabled, { color: 'grey' }),
          $ifelse<T>(disabled, { backgroundColor: 'blue' }, { backgroundColor: 'green' }),
        ))
      inRender(true)
      inRender(false)
    })
  }

  describe("## NATIVE ##", () => doTest(false))
  describe("## WEB ##", () => doTest(true))

})
