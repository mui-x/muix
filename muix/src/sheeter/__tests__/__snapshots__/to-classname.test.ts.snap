// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TO CLASSNAMES ## NATIVE ## 01: null 1`] = `null`;

exports[`TO CLASSNAMES ## NATIVE ## 01: null 2`] = `null`;

exports[`TO CLASSNAMES ## NATIVE ## 01: null 3`] = `undefined`;

exports[`TO CLASSNAMES ## NATIVE ## 02: null, {}, undefined 1`] = `null`;

exports[`TO CLASSNAMES ## NATIVE ## 02: null, {}, undefined 2`] = `null`;

exports[`TO CLASSNAMES ## NATIVE ## 02: null, {}, undefined 3`] = `undefined`;

exports[`TO CLASSNAMES ## NATIVE ## 03: mixed 1`] = `
Array [
  Array [
    "color: red /*.*/",
  ],
  Array [
    "color: blue /*root[0]*/",
  ],
  Array [
    "color: yellow /*root[1]*/",
  ],
  Array [
    "color: green /*.*/",
  ],
]
`;

exports[`TO CLASSNAMES ## NATIVE ## 03: mixed 2`] = `
Object {
  "color": Object {
    "tracePath": ".",
    "value": "green",
  },
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 03: mixed 3`] = `
Object {
  "color": "green",
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 04: query => opened 1`] = `
Array [
  Array [
    "color: red /*./$switch.isOpened*/",
  ],
]
`;

exports[`TO CLASSNAMES ## NATIVE ## 04: query => opened 2`] = `
Object {
  "color": Object {
    "tracePath": "./$switch.isOpened",
    "value": "red",
  },
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 04: query => opened 3`] = `
Object {
  "color": "red",
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 05: query => closed 1`] = `
Array [
  Array [
    "color: blue /*./$switch.isClosed*/",
  ],
]
`;

exports[`TO CLASSNAMES ## NATIVE ## 05: query => closed 2`] = `
Object {
  "color": Object {
    "tracePath": "./$switch.isClosed",
    "value": "blue",
  },
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 05: query => closed 3`] = `
Object {
  "color": "blue",
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 06: query => closed => pre-atomized 1`] = `
Array [
  Array [
    "color: red /*./$switch.isOpened*/",
  ],
  Array [
    "color: blue /*./$switch.isClosed*/",
  ],
]
`;

exports[`TO CLASSNAMES ## NATIVE ## 06: query => closed => pre-atomized 2`] = `
Object {
  "color": Object {
    "tracePath": "./$switch.isClosed",
    "value": "blue",
  },
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 06: query => closed => pre-atomized 3`] = `
Object {
  "color": "blue",
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 07: query => closed => pre-queried 1`] = `
Array [
  Array [
    "color: blue /*./$switch.isClosed*/",
  ],
]
`;

exports[`TO CLASSNAMES ## NATIVE ## 07: query => closed => pre-queried 2`] = `
Object {
  "color": Object {
    "tracePath": "./$switch.isClosed",
    "value": "blue",
  },
}
`;

exports[`TO CLASSNAMES ## NATIVE ## 07: query => closed => pre-queried 3`] = `
Object {
  "color": "blue",
}
`;

exports[`TO CLASSNAMES ## WEB ## 01: null 1`] = `null`;

exports[`TO CLASSNAMES ## WEB ## 01: null 2`] = `null`;

exports[`TO CLASSNAMES ## WEB ## 01: null 3`] = `undefined`;

exports[`TO CLASSNAMES ## WEB ## 02: null, {}, undefined 1`] = `null`;

exports[`TO CLASSNAMES ## WEB ## 02: null, {}, undefined 2`] = `null`;

exports[`TO CLASSNAMES ## WEB ## 02: null, {}, undefined 3`] = `undefined`;

exports[`TO CLASSNAMES ## WEB ## 03: mixed 1`] = `
Array [
  Array [
    ".d { color:red /*.*/ }",
  ],
  Array [
    ".a { color:blue /*root[0]*/ }",
  ],
  Array [
    ".b { color:yellow /*root[1]*/ }",
  ],
  Array [
    ".c { color:green /*.*/ }",
  ],
]
`;

exports[`TO CLASSNAMES ## WEB ## 03: mixed 2`] = `
Array [
  ".c { color:green /*.*/ }",
]
`;

exports[`TO CLASSNAMES ## WEB ## 03: mixed 3`] = `"c"`;

exports[`TO CLASSNAMES ## WEB ## 04: query => opened 1`] = `
Array [
  Array [
    ".a { color:red /*./$switch.isOpened*/ }",
  ],
]
`;

exports[`TO CLASSNAMES ## WEB ## 04: query => opened 2`] = `
Array [
  ".a { color:red /*./$switch.isOpened*/ }",
]
`;

exports[`TO CLASSNAMES ## WEB ## 04: query => opened 3`] = `"a"`;

exports[`TO CLASSNAMES ## WEB ## 05: query => closed 1`] = `
Array [
  Array [
    ".b { color:blue /*./$switch.isClosed*/ }",
  ],
]
`;

exports[`TO CLASSNAMES ## WEB ## 05: query => closed 2`] = `
Array [
  ".b { color:blue /*./$switch.isClosed*/ }",
]
`;

exports[`TO CLASSNAMES ## WEB ## 05: query => closed 3`] = `"b"`;

exports[`TO CLASSNAMES ## WEB ## 06: query => closed => pre-atomized 1`] = `
Array [
  Array [
    ".a { color:red /*./$switch.isOpened*/ }",
  ],
  Array [
    ".b { color:blue /*./$switch.isClosed*/ }",
  ],
]
`;

exports[`TO CLASSNAMES ## WEB ## 06: query => closed => pre-atomized 2`] = `
Array [
  ".b { color:blue /*./$switch.isClosed*/ }",
]
`;

exports[`TO CLASSNAMES ## WEB ## 06: query => closed => pre-atomized 3`] = `"b"`;

exports[`TO CLASSNAMES ## WEB ## 07: query => closed => pre-queried 1`] = `
Array [
  Array [
    ".b { color:blue /*./$switch.isClosed*/ }",
  ],
]
`;

exports[`TO CLASSNAMES ## WEB ## 07: query => closed => pre-queried 2`] = `
Array [
  ".b { color:blue /*./$switch.isClosed*/ }",
]
`;

exports[`TO CLASSNAMES ## WEB ## 07: query => closed => pre-queried 3`] = `"b"`;
